File: leetcode/43/43.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        unordered_map<char, string> storageAns;
        if(num1 == "0" || num2 == "0") 
            return "0";
        int k = 0;
        string partAns;
        string ans = "";
        for(int i = num2.size() - 1; i >= 0; i--, k++) {
            partAns = "";
            if(storageAns[num2[i]].size() != 0) {
                partAns = storageAns[num2[i]];
            }
            else if(num2[i] == '0')
                continue;
            else {
                for(char j = '0'; j < num2[i]; j++) {
                    partAns = addTwoString(partAns, num1);
                }
                storageAns[num2[i]] = partAns;
            }
            for(int j = 0; j < k; j++)
                partAns += '0';
            ans = addTwoString(ans, partAns);
        }
        return ans;
    }
    string addTwoString(string num1, string num2) {
        int i = num1.size() - 1;
        int j = num2.size() - 1;
        int flag = 0;  //用于表示两数相加是否有进位
        int integer1, integer2;
        string ans = "";
        while(i >= 0 && j >= 0) {
            integer1 = num1[i] - '0';
            integer2 = num2[j] - '0';
            ans = to_string((integer1 + integer2 + flag) % 10) + ans;
            flag = (integer1 + integer2 + flag) / 10;
            i--;
            j--;
        }
        while(i >= 0) {
            integer1 = num1[i] - '0';
            ans = to_string((integer1 + flag) % 10) + ans;
            flag = (integer1 + flag) / 10;
            i--;
        }
        while(j >= 0) {
            integer2 = num2[j] - '0';
            ans = to_string((integer2 + flag) % 10) + ans;
            flag = (integer2 + flag) / 10;
            j--;
        }
        if(flag) {
            ans = "1" + ans;
        }
        return ans;
    }
};

File: leetcode/43/43.java
class Solution {
    public String multiply(String num1, String num2) {
        String ans = "0";
        int len1 = num1.length(), len2 = num2.length();
        if(len1 == 0 || len2 == 0) {
            return "0";
        }

        if(num1.charAt(0) == '0' || num2.charAt(0) == '0') {
            return "0";
        }

        for(int i = len1 - 1; i >= 0; i--) {
            StringBuilder partSum = multiplyOneNumber(num2, num1.charAt(i) - '0');
            if(partSum.length() != 0) {
                for(int j = 0; j < len1 - i - 1; j++) {
                    partSum.append(0);
                }
                ans = sum(partSum.toString(), ans);
            }
        }
        return ans;
    }

    private StringBuilder multiplyOneNumber(String num, int oneNumber) {
        if(oneNumber == 0) {
            return new StringBuilder();
        }
        StringBuilder sb = new StringBuilder();
        int flag = 0;
        for(int i = num.length() - 1; i >= 0; i--) {
            int val = (num.charAt(i) - '0') * oneNumber + flag;
            flag = 0;
            if(val >= 10) {
                flag = val / 10;
                val = val % 10;
            }
            sb.append(val);
        }
        if(flag > 0) {
            sb.append(flag);
        }

        return sb.reverse();
    }

    private String sum(String num1, String num2) {
        int len1 = num1.length(), len2 = num2.length();
        if(len1 == 0) {
            return num2;
        }

        if(len2 == 0) {
            return num1;
        }

        StringBuilder sb = new StringBuilder();
        int flag = 0, i = len1 - 1, j = len2 - 1;
        for(; i >= 0 && j >= 0; i--, j--) {
            int val = (num1.charAt(i) - '0') + (num2.charAt(j) - '0') + flag;
            flag = 0;
            if(val >= 10) {
                flag = 1;
                val %= 10;
            }
            sb.append(val);
        }
        while(i >= 0) {
            int val = num1.charAt(i--) - '0' + flag;
            flag = 0;
            if(val >= 10) {
                flag = 1;
                val %= 10;
            }
            sb.append(val);
        }

        while(j >= 0) {
            int val = num1.charAt(j--) - '0' + flag;
            flag = 0;
            if(val >= 10) {
                flag = 1;
                val %= 10;
            }
            sb.append(val);
        }
        if(flag > 0) {
            sb.append(1);
        }

        return sb.reverse().toString();
    }
}

